### PR TITLE
CSVWriter: Performance enhancements

### PR DESCRIPTION
**Pull request by shanielh**
Hi mate, Great library.

I've got some performance issues when writing strings, I narrowed it to the library, so I did some changes : 

1. When `QUOTE_MINIMAL` is used, don't create a list for every field quote
2. Use  `+`  operator to quote field
3. Use a while loop to print fields instead of `mkstring`

### CODE REVIEW

File: src/main/scala/com/github/tototoshi/csv/CSVWriter.scala [hunk 1]
```diff
@@ -31,27 +32,26 @@ class CSVWriter(protected val writer: Writer)(implicit val format: CSVFormat) ex
...
             }
           }
-        }
       }
 
     def quote(field: String): String =
-      if (shouldQuote(field, format.quoting)) field.mkString(format.quoteChar.toString, "", format.quoteChar.toString)
-      else field
+      if (shouldQuote(field, format.quoting))
+        format.quoteChar + field + format.quoteChar.toString
```
Comment (tototoshi) [inline src/main/scala/com/github/tototoshi/csv/CSVWriter.scala]: `toString` is unnecessary?

**Comment by tototoshi** [comment]
Thanks. Do you have any benchmarking result?

**Comment by shanielh** [comment]
Benchmarks : 

###### Old 

```
cores: 4
hostname: shani-laptop
name: Java HotSpot(TM) 64-Bit Server VM
osArch: amd64
osName: Linux
vendor: Oracle Corporation
version: 25.92-b14
Parameters(quoting -> QUOTE_ALL, size -> 10000): 26.40353
Parameters(quoting -> QUOTE_ALL, size -> 100000): 273.820936
Parameters(quoting -> QUOTE_MINIMAL, size -> 10000): 30.928586
Parameters(quoting -> QUOTE_MINIMAL, size -> 100000): 314.662758
Parameters(quoting -> QUOTE_NONE, size -> 10000): 15.015368
Parameters(quoting -> QUOTE_NONE, size -> 100000): 152.473534
Parameters(quoting -> QUOTE_NONNUMERIC, size -> 10000): 24.142457
Parameters(quoting -> QUOTE_NONNUMERIC, size -> 100000): 247.681169
```

###### New 

```
cores: 4
hostname: shani-laptop
name: Java HotSpot(TM) 64-Bit Server VM
osArch: amd64
osName: Linux
vendor: Oracle Corporation
version: 25.92-b14
Parameters(quoting -> QUOTE_ALL, size -> 10000): 17.757893
Parameters(quoting -> QUOTE_ALL, size -> 100000): 183.255968
Parameters(quoting -> QUOTE_MINIMAL, size -> 10000): 17.596785
Parameters(quoting -> QUOTE_MINIMAL, size -> 100000): 182.140503
Parameters(quoting -> QUOTE_NONE, size -> 10000): 14.648283
Parameters(quoting -> QUOTE_NONE, size -> 100000): 149.096122
Parameters(quoting -> QUOTE_NONNUMERIC, size -> 10000): 19.521622
Parameters(quoting -> QUOTE_NONNUMERIC, size -> 100000): 200.327575
```

File: src/main/scala/com/github/tototoshi/csv/CSVWriter.scala [hunk 2]
```diff
@@ -69,7 +69,16 @@ class CSVWriter(protected val writer: Writer)(implicit val format: CSVFormat) ex
       quote _ compose escape compose show
     }
 
-    printWriter.print(fields.map(renderField).mkString(format.delimiter.toString))
+    val iterator = fields.iterator
+    var hasNext = iterator.hasNext
+    while (hasNext) {
+      printWriter.print(renderField(iterator.next()))
```
Comment (shanielh) [inline src/main/scala/com/github/tototoshi/csv/CSVWriter.scala]: It just came to me that we shouldn't render field to create a new string at all. We can write the rendered field directly to output. Should be faster and allocate less memory :)

**Comment by shanielh** [comment]
Hi, I fixed some bug in the code and added another performance enhancement, The code doesn't create a new string when it needs to quote a string (When the quote string doesn't contain the quote char)

The new benchmark results are : 

```
cores: 4
hostname: shani-laptop
name: Java HotSpot(TM) 64-Bit Server VM
osArch: amd64
osName: Linux
vendor: Oracle Corporation
version: 25.92-b14
Parameters(quoting -> QUOTE_ALL, size -> 10000): 14.437709
Parameters(quoting -> QUOTE_ALL, size -> 100000): 146.714325
Parameters(quoting -> QUOTE_MINIMAL, size -> 10000): 8.998118
Parameters(quoting -> QUOTE_MINIMAL, size -> 100000): 91.24499
Parameters(quoting -> QUOTE_NONE, size -> 10000): 3.42072
Parameters(quoting -> QUOTE_NONE, size -> 100000): 36.226731
Parameters(quoting -> QUOTE_NONNUMERIC, size -> 10000): 13.586739
Parameters(quoting -> QUOTE_NONNUMERIC, size -> 100000): 138.25522
```

**Comment by tototoshi** [comment]
LGTM 👍

**Pull request merged**